plugins/fileservices/fileservices.cpp

```diff
@@ -198,63 +198,129 @@ static IPropertyTree *getEnvironmentTree(IConstEnvironment * daliEnv)
     return getHPCCEnvironment();
 }
 
-static StringAttr espurl;           // Default ESP url if none specified
+static void setServerAccess(CClientFileSpray &server, IConstWorkUnit * wu)
+{
+    StringBuffer user, password, token;
+    wu->getSecurityToken(StringBufferAdaptor(token));
+    extractToken(token.str(), wu->queryWuid(), StringBufferAdaptor(user), StringBufferAdaptor(password));
+    server.setUsernameToken(user.str(), password.str(), "");
+}
+
+static StringArray availableWsFS;
 static CriticalSection espURLcrit;
 
-static const char *getEspServerURL(const char *param)
+static void addConfiguredWsFSUrl(const char * url)
+{
+    CriticalBlock b(espURLcrit);
+    availableWsFS.appendUniq(url);
+}
+
+static bool contactWsFS(const char * espurl, IConstWorkUnit * wu)
+{
+    CClientFileSpray server;
+    server.addServiceUrl(espurl);
+    setServerAccess(server, wu);
+
+    try
+    {
+        Owned<IClientEchoDateTime> req = server.createEchoDateTimeRequest();
+        Owned<IClientEchoDateTimeResponse> result = server.EchoDateTime(req);
+
+        if (result->getResult())
+            return true;
+    }
+    catch(IException *ie)
+    {
+        StringBuffer error;
+        ie->errorMessage(error);
+        PROGLOG("Could not contact WsFS: '%s': %s",espurl, error.str());
+        ie->Release();
+    }
+    catch(...)
+    {
+        PROGLOG("Could not contact WsFS: '%s'",espurl);
+    }
+    return false;
+}
+
+static const char * getNextAliveWsFSURL(IConstWorkUnit * wu)
+{
+    CriticalBlock b(espURLcrit);
+
+    for (int index = 0; index < availableWsFS.length(); index++)
+    {
+        const char * currentUrl = availableWsFS.item(index);
+        if (contactWsFS(currentUrl, wu))
+            return currentUrl;
+    }
+
+    return nullptr;
+}
+
+static bool isUrlListEmpty()
+{
+    CriticalBlock b(espURLcrit);
+    return availableWsFS.length() == 0;
+}
+
+static const char *getAccessibleEspServerURL(const char *param, IConstWorkUnit * wu)
 {
     if (param&&*param)
         return param;
 
     CriticalBlock b(espURLcrit);
-    if (espurl.isEmpty())
+    if (isUrlListEmpty())
     {
         Owned<IConstEnvironment> daliEnv = openDaliEnvironment();
         Owned<IPropertyTree> env = getEnvironmentTree(daliEnv);
-        StringBuffer tmp;
+        StringBuffer wsfsurl;
         if (env.get()) {
             Owned<IPropertyTreeIterator> iter1 = env->getElements("Software/EspProcess");
-            ForEach(*iter1) {
+            ForEach(*iter1)
+            {
                 Owned<IPropertyTreeIterator> iter2 = iter1->query().getElements("EspBinding");
-                ForEach(*iter2) {
+                ForEach(*iter2)
+                {
                     Owned<IPropertyTreeIterator> iter3 = iter2->query().getElements("AuthenticateFeature");
-                    ForEach(*iter3) {
+                    ForEach(*iter3)
```

Comment:
Agreed on all these points, and your suggestion is the obvious cleanest implementation. But no matter which way we go, we're making assumptions (WsFS will always be part of ecwatch, wsfs auth always associated with wsfs). It does't seem worthwhile to change the legacy implementation, but I'll go ahead a change it.